### ISSUE TITLE
DPT-ViT with FP16, GN and grad accum causes NaN loss.

### ISSUE THREAD

**Issue by lorinczszabolcs** [open-mmlab/mmsegmentation#1905]
**Checklist**

1. I have searched related issues but cannot get the expected help. ✅
2. The bug has not been fixed in the latest version. ✅

**Describe the bug**
When using DPT-ViT with `FP16`. gradient accumulation for 4 iterations and `GN` instead of `SyncBN` (single GPU setting, using `GN` due to small `batch_size=4` with `num_groups=8`), the loss becomes NaN after some iterations.

**Reproduction**

1. What command or script did you run?
I am using a custom training script, but basically first calling `build_dataset`, then `build_segmentor`, finally `train_segmentor`.

2. Did you make any modifications on the code or config? Did you understand what you have modified?
Yes, extended the original implementation to enable having multiple heads to predict multiple targets, apart from that I am using the original `ViT` + `DPTHead` implementations, the heads are just simply duplicated.

Also changed 
```python
# Group Normalization instead of SyncBN due to single GPU training and small batch size
norm_cfg = dict(type="GN", num_groups=8, requires_grad=True)
```

and

```python
# batch size: 4
data = dict(samples_per_gpu=4, workers_per_gpu=4)

# cumulative_iters: 4 -> real batch size: 4 x 4 = 16
optimizer_config = dict(
    type="GradientCumulativeFp16OptimizerHook", cumulative_iters=4, loss_scale="dynamic"
)

# fp16 placeholder
fp16 = {}
```

3. What dataset did you use?
Custom data set, having multiple targets, one per head.

**Environment**

1. Please run `python mmseg/utils/collect_env.py` to collect necessary environment information and paste it here.
```
sys.platform: linux
Python: 3.8.12 (default, Oct 12 2021, 13:49:34) [GCC 7.5.0]
CUDA available: True
GPU 0: NVIDIA GeForce RTX 3090
CUDA_HOME: None
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.11.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.5.2 (Git Hash a9302535553c73243c632ad3c4c80beec3d19a1e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.11.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.12.0
OpenCV: 4.6.0
MMCV: 1.5.2
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.3
MMSegmentation: 0.25.0+cf6ddbd
```
3. You may add addition that may be helpful for locating the problem, such as
   - PyTorch is running in a docker container, based on the latest PyTorch image, no other env variables might be related in my opinion.

**NOTE**

When turning `FP16` off via replacing the original config with
```python
# batch size: 4
data = dict(samples_per_gpu=4, workers_per_gpu=4)

# cumulative_iters: 4 -> real batch size: 4 x 4 = 16
optimizer_config = dict(type="GradientCumulativeOptimizerHook", cumulative_iters=4)
```
the NaN loss values disappear. All the other settings (`GN` and grad accum remain the same).

**Comment by MeowZheng** [comment]
I think FP16 might not apply to any cases, as there are always round errors when casting  FP32 to FP16.

**Comment by lorinczszabolcs** [comment]
Hi! Thanks for the quick response. Do you mean by that, that it might happen sometimes with FP16 and it is expected as such?

**Comment by MeowZheng** [comment]
Sorry for typo. It is roundoff error or rounding error, as the precision and range of fp16 is lower than fp32.

**Comment by lorinczszabolcs** [comment]
Ok, I see. Thank you!

**Comment by bestfleer** [comment]
can you tell me how to use bfloat16 in mmseg? bfloat16 has the same range as fp32 and can solve this problem

> Sorry for typo. It is roundoff error or rounding error, as the precision and range of fp16 is lower than fp32.
